pkg/controller/user-cluster-controller-manager/ccm-csi-migrator/controller.go

```diff
@@ -0,0 +1,189 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ccmcsimigrator
+
+import (
+	"context"
+	"fmt"
+	"github.com/kubermatic/machine-controller/pkg/apis/cluster/common"
+	"github.com/kubermatic/machine-controller/pkg/apis/cluster/v1alpha1"
+	"go.uber.org/zap"
+	v1 "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1"
+	"k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1/helper"
+	"k8c.io/kubermatic/v2/pkg/version/kubermatic"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"strconv"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/record"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerName = "ccm-csi-migrator"
+)
+
+type reconciler struct {
+	log          *zap.SugaredLogger
+	seedClient   ctrlruntimeclient.Client
+	userClient   ctrlruntimeclient.Client
+	seedRecorder record.EventRecorder
+	versions     kubermatic.Versions
+	clusterName  string
+}
+
+func Add(ctx context.Context, log *zap.SugaredLogger, seedMgr, userMgr manager.Manager, versions kubermatic.Versions, clusterName string) error {
+	log = log.Named(controllerName)
+
+	r := &reconciler{
+		log:          log,
+		seedClient:   seedMgr.GetClient(),
+		userClient:   userMgr.GetClient(),
+		seedRecorder: seedMgr.GetEventRecorderFor(controllerName),
+		versions:     versions,
+		clusterName:  clusterName,
+	}
+	c, err := controller.New(controllerName, userMgr, controller.Options{
+		Reconciler: r,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create controller: %v", err)
+	}
+
+	commonHandler := func(obj ctrlruntimeclient.Object, queue workqueue.RateLimitingInterface) {
+		cluster := &v1.Cluster{}
+		if err := r.seedClient.Get(ctx, types.NamespacedName{
+			Name: r.clusterName,
+		}, cluster); err != nil {
+			log.Warnf("cluster %s not found", clusterName)
+			return
+		}
+
+		// consider only clusters needing the migration
+		_, ccmOk := cluster.Annotations[v1.CCMMigrationNeededAnnotation]
+		_, csiOk := cluster.Annotations[v1.CSIMigrationNeededAnnotation]
+		if ccmOk && csiOk {
+			queue.Add(reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name: clusterName,
+				},
+			})
+		}
+	}
+
+	// Watch for changes to Machines
+	if err = c.Watch(
+		&source.Kind{Type: &v1alpha1.Machine{}},
+		&handler.Funcs{
+			CreateFunc: func(e event.CreateEvent, queue workqueue.RateLimitingInterface) { commonHandler(e.Object, queue) },
```

Comment:
Since we moved to the ccm-migration flag mechanism, I removed the filter (if the controller runs, it necessarily has to reconcile the machines for the given cluster)